### ISSUE TITLE
[Datagrid] Feature: First party scroll (performant) restoration

### ISSUE THREAD

**Issue by lauri865** [mui/mui-x#15190]
### Search keywords

restore scroll

### Latest version

- [X] I have tested the latest version

### Summary

We implement most of our Datagrids while persisting state between navigations for better UX. It's all good implemented in userland, and no complaints there in general.

However, restoring scroll is extremely painful in userland, and it's impossible to achieve a performant outcome there, because the initial renderContext is flushSynced, and there's no way to override that. So you end up with reflows and unnecessary component renders (even if they don't end up in the DOM, e.g. canvas elements end up still be drawn and have to be trashed if they don't end up in the viewport) on remounting the grid.

### Examples

What works but is slow / subject to reflows:
```typescript
apiRef.current.subscribeEvent("viewportInnerSizeChange", ...
// 1. Set up MutationObserver
// 2. Check for whether `.MuiDataGrid-virtualScroller` exists
// 3. Restore scroll
```

Slightly more performant, but less widely supported option (based on heuristics rather than explictly waiting for the virtualScroll to enter the DOM):
```typescript
React.useLayoutEffect(() => {
  queueMicrotask(() => {
      const rootRef = apiRef.current.rootElementRef?.current;
      const scrollContainer = rootRef?.querySelector(
        ".MuiDataGrid-virtualScroller",
      );
      if (!scrollContainer) return;

      if (gridState.scroll!.top) {
        scrollContainer.scrollTop = gridState.scroll!.top;
      }
      if (gridState.scroll!.left) {
        scrollContainer.scrollLeft = gridState.scroll!.left;
      }
  });
}, []);
```

Even more performant option (almost there, but could be better):
```typescript
React.useLayoutEffect(() => {
  queueMicrotask(() => {
      const rootRef = apiRef.current.rootElementRef?.current;
      const scrollContainer = rootRef?.querySelector(
        ".MuiDataGrid-virtualScroller",
      );
      if (!scrollContainer) return;

      flushSync(() => {
        apiRef.current.setState((state) => ({
          ...state,
          virtualization: {
            ...state.virtualization,
            renderContext: gridState.scroll?.renderContext!,
          },
        }));
      });

      if (gridState.scroll!.top) {
        scrollContainer.scrollTop = gridState.scroll!.top;
      }
      if (gridState.scroll!.left) {
        scrollContainer.scrollLeft = gridState.scroll!.left;
      }
  });
}, []);
```

Almost as snappy as without scroll restoration, but expensive components are still called and flushSync needs to wait for another flushSync to complete from what I understand.

### Motivation

The source of the issue seems to be – the following call to `updateRenderContext` cannot be intercepted or overridden:
https://github.com/mui/mui-x/blob/f613377c4740cb7ebda19d4b3a81cf27be8e0a01/packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx#L580C3-L591C6

FlushSynced:
https://github.com/mui/mui-x/blob/f613377c4740cb7ebda19d4b3a81cf27be8e0a01/packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx#L279-L282

Maybe a stopgap solution could be if `scrollPositionChange` could prevent the default behaviour (e.g. could prevent the first event and first invocation of `updateRenderContext`).

Restoring scroll is difficult in userland, subject to performance woes and regardless of the approach, ends up in spaghetti code that may not work in future versions.

Related: #5071 and #6411
Opened a new issue as these discussions are stale and based on old versions.

**Comment by michelengelen** [comment]
Hey @lauri865 and thanks for opening an issue here.
Also thanks for providing so much information and potential use-cases to look into. As scrolling issues seem to be a very common thing in the last few days I will add this to the board for the team to have a look into it in more detail.

@mui/xgrid thoughts?

**Comment by michelengelen** [comment]
related issue: #15179

**Comment by lauri865** [comment]
As an illustration what happens:
First render (scrolled to the top): `12 components/rows render`
*scroll to the bottom, navigate away and back*
Restore scroll to the bottom:
```
[Violation] Forced reflow while executing JavaScript took 31ms
13 render
8 render
[Violation] Forced reflow while executing JavaScript took 37ms
3 render
```

It's even worse when the scroll is in the middle of the grid. Then it can render 13+23+3=39 rows (while the viewport fits 8-9) in order to display the restored scroll position.

**Comment by lauri865** [comment]
I found the most disgusting way to prevent the first slice form being rendered 🤡:
```typescript
  const unsubscribe = apiRef.current.subscribeEvent("stateChange", (data) => {
      data.virtualization.renderContext = gridState.renderContext;
  });
```

(need to unsubscribe from the event when the grid is ready / after restoring scroll)

You will probably get fired if you use it at work though, but it's the only option that gets rid of unnecessary rendering.

**Comment by piyushyadav1617** [comment]
Does it not work for community version?

**Comment by cherniavskii** [comment]
@piyushyadav1617 Yes, but it's not released yet. The release is planned for today.

**Comment by piyushyadav1617** [comment]
> [@piyushyadav1617](https://github.com/piyushyadav1617) Yes, but it's not released yet. The release is planned for today.

Thanks! 
So this feature is releasing today for community version as well?
